itest/lnd_misc_test.go

```diff
@@ -1308,3 +1308,105 @@ func testNativeSQLNoMigration(ht *lntest.HarnessTest) {
+	ht.FundCoinsNP2WKH(btcutil.SatoshiPerBitcoin, alice)
+	ht.FundCoinsP2TR(btcutil.SatoshiPerBitcoin, alice)
+
+	// Get all the utxos in the wallet and assert there are three.
+	utxos := ht.AssertNumUTXOs(alice, 3)
+
+	// Ensure that we can't send duplicate coins.
+	//
+	// Create duplciate outpoints.
+	dupOutpoints := []*lnrpc.OutPoint{
+		utxos[0].Outpoint,
+		utxos[0].Outpoint,
+	}
+
+	// Send the duplicate outpoints and assert there's an error.
+	err := alice.RPC.SendCoinsAssertErr(&lnrpc.SendCoinsRequest{
+		Addr:      ht.NewMinerAddress().String(),
+		Outpoints: dupOutpoints,
+	})
+	require.ErrorContains(ht, err, "selected outpoints contain duplicate")
+
+	// Send a selected coin with a specific amount.
+	//
+	// We'll send the first utxo with an amount of 0.5 BTC.
+	amt := btcutil.Amount(0.5 * btcutil.SatoshiPerBitcoin)
+	alice.RPC.SendCoins(&lnrpc.SendCoinsRequest{
+		Addr: ht.NewMinerAddress().String(),
+		Outpoints: []*lnrpc.OutPoint{
+			utxos[0].Outpoint,
+		},
+		Amount: int64(amt),
+	})
+
+	// We expect to see the above tx in the mempool.
+	tx := ht.GetNumTxsFromMempool(1)[0]
+
+	// Assert the tx has the expected shape. It should have 1 input and 2
+	// outputs - the input is the selected UTXO, and the outputs are the
+	// specified amount and the change amount.
+	require.Len(ht, tx.TxIn, 1)
+	require.Len(ht, tx.TxOut, 2)
+
+	// Check it's using the selected UTXO as input.
+	require.Equal(ht, utxos[0].Outpoint.TxidStr,
+		tx.TxIn[0].PreviousOutPoint.Hash.String())
+
+	// Mine a block to confirm the above tx.
+	ht.MineBlocksAndAssertNumTxes(1, 1)
+
+	// We now test we can send the selected coins to a single target
+	// address with `SendAll` flag.
+	//
+	// Get all the utxos in the wallet and assert there are three.
+	utxos = ht.AssertNumUTXOs(alice, 3)
+
+	// Select the first two coins and send them to a single target address.
+	alice.RPC.SendCoins(&lnrpc.SendCoinsRequest{
+		Addr: ht.NewMinerAddress().String(),
+		Outpoints: []*lnrpc.OutPoint{
+			utxos[0].Outpoint,
+			utxos[1].Outpoint,
+		},
+		SendAll: true,
+	})
+
+	// We expect to see the above tx in the mempool.
+	tx = ht.GetNumTxsFromMempool(1)[0]
+
+	// Assert the tx has the expected shape. It should have 2 inputs and 1
+	// output.
+	require.Len(ht, tx.TxIn, 2)
+	require.Len(ht, tx.TxOut, 1)
+
+	// Check it's using the selected UTXOs as inputs.
+	prevOutpoint1 := tx.TxIn[0].PreviousOutPoint.Hash.String()
+	prevOutpoint2 := tx.TxIn[1].PreviousOutPoint.Hash.String()
+
+	if prevOutpoint1 == utxos[0].Outpoint.TxidStr {
+		require.Equal(ht, utxos[1].Outpoint.TxidStr, prevOutpoint2)
+	} else {
+		require.Equal(ht, utxos[1].Outpoint.TxidStr, prevOutpoint1)
+		require.Equal(ht, utxos[0].Outpoint.TxidStr, prevOutpoint2)
+	}
+
+	// Mine a block to confirm the above tx.
+	ht.MineBlocksAndAssertNumTxes(1, 1)
+
+	// Since the first two UTXOs have been sent to an address outside her
+	// wallet, Alice should see a single UTXO now.
+	ht.AssertNumUTXOs(alice, 1)
```

Comment:
Could we further check that the remaining coin can't be selected and spent in its entirety so that the required reserves are not violated?

On this note, should the API allow to spend the coin up until the required reserve and not return an error unless other constraints like the dust limit of the sending amount are reached? Maybe further tests can document behavior here as well.

Reply:
yeah I like it, added a test to show how it behaves atm.